api/types/authentication_authpreference_test.go

```diff
@@ -331,3 +333,99 @@ func TestAuthPreferenceV2_CheckAndSetDefaults(t *testing.T) {
 		})
 	}
 }
+
+func TestAuthPreferenceV2_GetPreferredLocalMFA(t *testing.T) {
+	u2fConfig := &types.U2F{
+		AppID: "https://example.com",
+		Facets: []string{
+			"https://example.com:443",
+			"https://example.com",
+			"example.com:443",
+			"example.com",
+		},
+	}
+	webConfig := &types.Webauthn{
+		RPID: "example.com",
+	}
+	disabledWebConfig := &types.Webauthn{
+		RPID:     "example.com",
+		Disabled: true,
+	}
+
+	tests := []struct {
+		name string
+		spec types.AuthPreferenceSpecV2
+		want constants.SecondFactorType
+	}{
+		{
+			name: "sf=off",
+			spec: types.AuthPreferenceSpecV2{
+				Type:         constants.Local,
+				SecondFactor: constants.SecondFactorOff,
+			},
+		},
+		{
+			name: "sf=otp",
+			spec: types.AuthPreferenceSpecV2{
+				Type:         constants.Local,
+				SecondFactor: constants.SecondFactorOTP,
+			},
+			want: constants.SecondFactorOTP,
+		},
+		{
+			name: "sf=u2f",
+			spec: types.AuthPreferenceSpecV2{
+				Type:         constants.Local,
+				SecondFactor: constants.SecondFactorU2F,
+				U2F:          u2fConfig,
+			},
+			want: constants.SecondFactorU2F,
+		},
+		{
+			name: "sf=webauthn",
+			spec: types.AuthPreferenceSpecV2{
+				Type:         constants.Local,
+				SecondFactor: constants.SecondFactorWebauthn,
+				Webauthn:     webConfig,
+			},
+			want: constants.SecondFactorWebauthn,
+		},
+		{
+			name: "sf=on favors WebAuthn",
+			spec: types.AuthPreferenceSpecV2{
+				Type:         constants.Local,
+				SecondFactor: constants.SecondFactorOn,
+				U2F:          u2fConfig,
+				Webauthn:     webConfig,
+			},
+			want: constants.SecondFactorWebauthn,
+		},
+		{
+			name: "sf=on disabled WebAuthn favors U2F",
+			spec: types.AuthPreferenceSpecV2{
+				Type:         constants.Local,
+				SecondFactor: constants.SecondFactorOn,
+				U2F:          u2fConfig,
+				Webauthn:     disabledWebConfig,
+			},
+			want: constants.SecondFactorU2F,
+		},
+		{
+			name: "sf=optional favors WebAuthn",
+			spec: types.AuthPreferenceSpecV2{
+				Type:         constants.Local,
+				SecondFactor: constants.SecondFactorOptional,
+				U2F:          u2fConfig,
+				Webauthn:     webConfig,
+			},
+			want: constants.SecondFactorWebauthn,
+		},
```

Comment:
is it redundant to also test for `sf=optional disabled WebAuthn favors U2F`

Reply:
Sure, makes sense. Done.